subt_ign/src/world_generator_cave.cc

```diff
@@ -0,0 +1,1539 @@
+      // if we have not reached the desired tile count yet.
+      bool valid = !(openings.empty() &&
+          s.connectionPoints.empty() &&
+          tileCount - s.tiles.size() > 0);
+      if (!valid)
+        continue;
+
+      // set the world pose of the new section
+      s.pose = transform;
+
+      // if it is a transition tile, we need to make sure connection points
+      // match. The transition world section created has Type A connection
+      // point at +x and Type B connection point at -x
+      bool flip = false;
+      if (tileTypeToSelect == CAVE_TYPE_TRANSITION)
+      {
+        // if the tile to connect to is Type A, we need to flip the transition
+        // tile by 180 degrees
+        if (tileType == CAVE_TYPE_A)
+        {
+          auto &transitionTile = s.tiles[0].model;
+          transitionTile.SetPose(math::Pose3d(transitionTile.Pose().Pos(),
+              math::Quaterniond(0, 0, IGN_PI) * transitionTile.Pose().Rot()));
+          flip = true;
+        }
+      }
+
+      // add the new section
+      selected = true;
+      selectedSection = s;
+      addedWorldSections.push_back(selectedSection);
+
+      // add the openings of the new world section to the list
+      // be sure to convert to world frame
+      for (auto &cp : s.connectionPoints)
+      {
+        ConnectionOpening op;
+        op.pos = transform.CoordPositionAdd(cp.first);
+        op.rot = transform.Rot() * cp.second;
+
+        // mark the tile type for the opening of the transition tile
+        // based on the tile type the opening connects to.
+        if (s.tileType == CAVE_TYPE_TRANSITION)
+        {
+          if (flip)
+            op.tileType = CAVE_TYPE_B;
+          else
+            op.tileType = CAVE_TYPE_A;
+        }
+        else
+        {
+          op.tileType = s.tileType;
+        }
+        openings.push_back(op);
+      }
+
+      // update number of tiles that still need to be added
+      tileCount -= s.tiles.size();
+    }
+
+    // if we fail to add a section, add to the unfilled list
+    // so we can put a cap there to block the opening
+    if (attempt >= maxAttempt)
+    {
+      unfilledOpenings.push_back(o);
+    }
+  }
+
+  // add remaining openings to the unfilled list
+  for (auto &o : openings)
+    unfilledOpenings.push_back(o);
+
+  // begin generating the world sdf
+  std::stringstream ss;
+  math::AxisAlignedBox worldBBox;
+  int tileNo = 0;
+  int caveTypeACount = 0;
+  int caveTypeBCount = 0;
+  int caveTransitionCount = 0;
+  for (const auto &s : addedWorldSections)
+  {
+    for (const auto &t : s.tiles)
+    {
+      std::string name = "tile_" + std::to_string(++tileNo);
+
+      // convert tile pose to world coordinates
+      math::Pose3d pose = t.model.Pose() + s.pose;
+      std::string uri =
+          "https://fuel.ignitionrobotics.org/1.0/OpenRobotics/models/" +
+          t.tileType;
+
+      ss << "    <include>\n";
+      ss << "      <static>true</static>\n";
```

Comment:
Should all tiles be static?

Reply:
yes they should be static. The models are actually already marked as static in the model.sdf file so this is probably redundant. I added this to be consistent with the existing world files.